Data/ByteString/Lazy/ReadNat.hs

```diff
@@ -0,0 +1,254 @@
+readNatural bs = do
+    (w, s) <- uncons bs
+    let d = fromDigit w
+    if | d <=    9 -> Just $! _readDecimal s d 2 []
+       | otherwise -> Nothing
+
+----- Internal implementation
+
+-- | Intermediate result from scanning a chunk, final output is
+-- obtained via `convert` after all the chunks are processed.
+--
+data Result = Result !Int      -- Bytes consumed
+                     !Word     -- Value of LSW
+                     !Int      -- Digits in LSW
+                     [Natural] -- Little endian MSW list
+
+_readDecimal :: ByteString -> Word -> Int -> [Natural]
+             -> (Natural, ByteString)
+_readDecimal = consume
+  where
+    consume :: ByteString -> Word -> Int -> [Natural] -> (Natural, ByteString)
+#ifdef BYTESTRING_STRICT
+    consume (BI.BS fp len) acc _ _ =
+        -- Having read one digit, we're about to read the 2nd
+        -- So the digit count up to 'safeLog' starts at 2.
+        case natdigits fp len acc 2 [] of
+            Result used acc' cnt ns
+                | used == len
+                  -> convert acc' cnt ns $ empty
+                | otherwise
+                  -> convert acc' cnt ns $
+                     BI.BS (fp `BI.plusForeignPtr` used) (len - used)
+#else
+    -- All done
+    consume Empty acc cnt ns = convert acc cnt ns Empty
+    -- Process next chunk
+    consume (Chunk (BI.BS fp len) cs) acc cnt ns
+        = case natdigits fp len acc cnt ns of
+            Result used acc' cnt' ns'
+                | used == len -- process more chunks
+                  -> consume cs acc' cnt' ns'
+                | otherwise   -- ran into a non-digit
+                  -> convert acc' cnt' ns' $
+                     Chunk (BI.BS (fp `BI.plusForeignPtr` used) (len - used)) cs
+#endif
+    convert !acc !cnt !ns rest =
+        let !n = combine acc cnt ns
+         in (n, rest)
+
+    -- | Merge least-significant word with reduction of of little-endian tail.
+    --
+    -- The input is:
+    --
+    -- * Least significant digits as a 'Word' (LSW)
+    -- * The number digits that went into the LSW
+    -- * All the remaining digit groups ('safeLog' digits each),
+    --   in little-endian order
+    --
+    -- The result is obtained by pairwise recursive combining of all the
+    -- full size digit groups, followed by multiplication by @10^cnt@ and
+    -- addition of the LSW.
+    --
+    combine :: Word      -- ^ value of LSW
+            -> Int       -- ^ count of digits in LSW
+            -> [Natural] -- ^ tail elements (base @10^'safeLog'@)
+            -> Natural
+    {-# INLINE combine #-}
+    combine !acc !_   [] = wordToNatural acc
+    combine !acc !cnt ns =
+        wordToNatural (10^cnt) * combine1 safeBase ns + wordToNatural acc
+
+    -- | Recursive reduction of little-endian sequence of 'Natural'-valued
+    -- /digits/ in base @base@ (a power of 10).  The base is squared after
+    -- each round.  The overall complexity is roughly /O(n log(n))/.
```

Comment:
This complexity estimate is still wrong. In the case of n digits for n a power of 2, there is 1 multiplications of size n/2; 2 multiplications of size n/4; 4 multiplications of size n/8; and so on... which adds up to O(M(n) log(n)) under reasonable assumptions about the complexity of the underlying multiplication algorithm.

Reply:
Yes, but worst case all the multiplies are n-bit long.  And I don't need to be precise here, again not public contract, just a comment for maintainers.

Reply:
> Yes, but worst case all the multiplies are n-bit long.

This doesn't make sense to me. The whole point of the recursive reduction is to perform fewer large multiplications.


> And I don't need to be precise here, again not public contract, just a comment for maintainers.

I think the comment as-is is very misleading. The presence of the log(n) creates some illusion of precision, even with the word "roughly." It suggests the considerations it omits are less significant than the one log factor it includes, when they are not. Of course the set of people it could mislead is smaller since this in an internal comment, but I think that's a poor excuse not to use one of the several less wrong alternatives:
* subquadratic
* Õ(n)
* roughly O(n log(n)^2)
* O(n log(n)^2 log(log(n)))

Reply:
I guess one way to look at it is the total size of memory allocated and freed is no longer quadratic, as the product length grows incrementally 64 bits at a time.  Instead we touch `n log(n)` bytes of memory for the result.  The CPU cost of the multiplications is indeed a more complex matter, perhaps it is simply easier to say "more efficient" or sub-quadratic, ...

How detailed an explanation do you think is warranted here?

Reply:
Never mind, the memory allocation is quadratic... I guess I don't have a good explanation of exactly how this is faster, I'd need to better understand the performance costs of the underlying bignum multiplication for `bignum * word`, and `bignum * bignum`.  The MR originally had `bignum * 10 + digit`, that's clearly more costly than `bignum * 10^19 + word`, so we're ahead there, but it is less clear why the combine loops are faster than `bignum * 10^19 + word` would be.  I should perhaps try it and compare...

Reply:
Folding in each word as it crosses 19 digits gives a reduction in code size with no visible performance cost!  So it appears this was never about `O`-complexity, but rather about the constant factors.  The recursive reduction from master does not appear to be a win, just the chunking is worthwhile...

Reply:
Shall I push fixup along these lines?

Reply:
> Never mind, the memory allocation is quadratic...

There are log(n) passes of `combine2`, each of which creates `Natural`s with sizes totaling about n bits, right? Each pass of `combine2` cuts the length of the current list in half. The difference in runtime asymptotics takes a while to become apparent, though, because multiplication of two large `Natural`s doesn't start to become obviously subquadratic until they are a couple-dozen words long each, which is a few hundred decimal digits. The "ReadNatural" benchmark's largest number is 2^(2^13), which at 2467 digits long only moderately benefits from these large multiplications, while it's also carrying around the overhead of building these lists. (It's still about 20% faster on my machine, though.) But when I add to your benchmarks a "ReadInteger huge" which is the same as the existing "ReadInteger" in this MR but with 2^2^20 (which has about 316K digits) as its limit instead of 2^2^13, I measure 55 ms ± 465 μs for 8d2e919 and 1.30 s ± 3.6 ms for e654544, which is a very large difference.

Reply:
Yes, for `2^2^20` I can reproduce performance degradation with the more concise code, I see a factor of ~10x (30ms -> 300ms). For some reason your numbers come out noticeably worse. Could it be GHC 9.2 vs. some other version?

Anyway, does dealing with numbers much larger than are seen in applied crypt warrant the fancier code? I can certainly drop the most recent commit, and then we'll need to agree what to say about the performance claims.  Can you suggest specific text?

Thanks for sticking out this far by the way.  Much appreciated.